File: csv-to-json-converter.js
// CSV to BookTracker JSON Converter
const fs = require('fs');
const path = require('path');

function csvToJson(csvPath, outputPath) {
    try {
        // Read the CSV file
        const csvContent = fs.readFileSync(csvPath, 'utf-8');
        const lines = csvContent.trim().split('\n');
        
        // Parse header
        const headers = lines[0].split(',').map(header => header.trim());
        console.log('CSV Headers:', headers);
        
        // Initialize books array
        const books = [];
        
        // Process each data row
        for (let i = 1; i < lines.length; i++) {
            const values = parseCSVLine(lines[i]);
            
            if (values.length < headers.length) {
                console.log(`Skipping incomplete row ${i}: ${lines[i]}`);
                continue;
            }
            
            // Map CSV data to BookTracker format
            const book = {
                id: generateId(),
                title: cleanValue(values[5]) || 'Unknown Title', // Title
                author: cleanValue(values[2]) || 'Unknown Author', // Author
                series: cleanValue(values[3]) || '', // Series Title
                bookNumber: parseNumber(values[4]) || null, // #In Series
                status: determineStatus(values[0], values[1]), // Read, DNF
                rating: parseRating(values[15]) || 0, // Rating
                mythicalElement: '', // Not in CSV
                publisher: cleanValue(values[16]) || '', // Publisher
                pageCount: parseNumber(values[6]) || null, // Page Count
                coverImage: '', // Not in CSV
                edition: cleanValue(values[7]) || '', // Edition
                
                // Physical features
                digitallySigned: parseBoolean(values[8]), // Digitally Signed
                signed: parseBoolean(values[9]), // Signed
                sprayedEdges: parseBoolean(values[10]), // Sprayed Edges
                hiddenCover: parseBoolean(values[11]), // Hidden Cover
                reversibleDustJacket: parseBoolean(values[12]), // Reversable Dust Jacket (note: misspelled in CSV)
                
                // Reading dates
                startedReading: parseDate(values[13]) || '', // Started Reading
                finishedReading: parseDate(values[14]) || '', // Finished Reading
                
                // Other fields
                gifted: parseGifted(values[17]), // Gifted
                notes: '', // Not in CSV
                dateAdded: new Date().toISOString()
            };
            
            books.push(book);
        }
        
        // Write to JSON file
        fs.writeFileSync(outputPath, JSON.stringify(books, null, 2));
        console.log(`Successfully converted ${books.length} books to ${outputPath}`);
        
        return books;
        
    } catch (error) {
        console.error('Error converting CSV to JSON:', error);
        throw error;
    }
}

// Helper functions
function parseCSVLine(line) {
    const values = [];
    let current = '';
    let inQuotes = false;
    
    for (let i = 0; i < line.length; i++) {
        const char = line[i];
        
        if (char === '"' && (i === 0 || line[i-1] === ',')) {
            inQuotes = true;
        } else if (char === '"' && inQuotes && (i === line.length - 1 || line[i+1] === ',')) {
            inQuotes = false;
        } else if (char === ',' && !inQuotes) {
            values.push(current);
            current = '';
        } else {
            current += char;
        }
    }
    values.push(current); // Add the last value
    
    return values;
}

function cleanValue(value) {
    if (!value) return '';
    return value.trim().replace(/^"(.*)"$/, '$1'); // Remove surrounding quotes
}

function parseNumber(value) {
    const cleaned = cleanValue(value);
    const num = parseInt(cleaned, 10);
    return isNaN(num) ? null : num;
}

function parseBoolean(value) {
    const cleaned = cleanValue(value).toUpperCase();
    return cleaned === 'TRUE' || cleaned === '1' || cleaned === 'YES';
}

function parseRating(value) {
    const cleaned = cleanValue(value);
    if (!cleaned) return 0;
    
    // Handle formats like "5/5", "4.5/5", "4"
    if (cleaned.includes('/')) {
        const parts = cleaned.split('/');
        const rating = parseFloat(parts[0]);
        return isNaN(rating) ? 0 : rating;
    }
    
    const rating = parseFloat(cleaned);
    return isNaN(rating) ? 0 : rating;
}

function parseDate(value) {
    const cleaned = cleanValue(value);
    if (!cleaned) return '';
    
    // Handle DD/MM/YYYY format
    if (cleaned.includes('/')) {
        const parts = cleaned.split('/');
        if (parts.length === 3) {
            const day = parts[0].padStart(2, '0');
            const month = parts[1].padStart(2, '0');
            const year = parts[2];
            return `${year}-${month}-${day}`;
        }
    }
    
    return cleaned;
}

function parseGifted(value) {
    const cleaned = cleanValue(value);
    // If there's a name in the Gifted column, it means it was gifted
    return cleaned !== '' && cleaned.toLowerCase() !== 'false';
}

function determineStatus(readValue, dnfValue) {
    const read = parseBoolean(readValue);
    const dnf = parseBoolean(dnfValue);
    
    if (dnf) return 'dnf';
    if (read) return 'read';
    return 'want-to-read';
}

function generateId() {
    return Date.now().toString(36) + Math.random().toString(36).substr(2);
}

// Main execution
const csvPath = '/Users/crazybuy/Downloads/Copy of Booklist - Sheet1.csv';
const outputPath = '/Users/crazybuy/Book Website/imported-books.json';

try {
    const books = csvToJson(csvPath, outputPath);
    console.log('\nConversion completed successfully!');
    console.log(`Total books converted: ${books.length}`);
    console.log(`Output file: ${outputPath}`);
    
    // Show some sample data
    console.log('\nSample books:');
    books.slice(0, 3).forEach((book, index) => {
        console.log(`\n${index + 1}. ${book.title} by ${book.author}`);
        console.log(`   Status: ${book.status}, Rating: ${book.rating}, Pages: ${book.pageCount}`);
        console.log(`   Physical features: ${Object.entries({
            'Digitally Signed': book.digitallySigned,
            'Signed': book.signed, 
            'Sprayed Edges': book.sprayedEdges,
            'Hidden Cover': book.hiddenCover,
            'Reversible Dust Jacket': book.reversibleDustJacket
        }).filter(([key, value]) => value).map(([key]) => key).join(', ') || 'None'}`);
    });
    
} catch (error) {
    console.error('Failed to convert CSV:', error);
    process.exit(1);
}

File: csv_to_json_converter.py
#!/usr/bin/env python3
import csv
import json
import time
import random
from datetime import datetime

def generate_id():
    """Generate a unique ID similar to the JavaScript version"""
    timestamp = int(time.time() * 1000)
    random_part = ''.join(random.choices('abcdefghijklmnopqrstuvwxyz0123456789', k=8))
    return f"{timestamp:x}{random_part}"

def clean_value(value):
    """Clean and normalize CSV values"""
    if not value:
        return ''
    return str(value).strip().strip('"')

def parse_number(value):
    """Parse a string to integer, return None if invalid"""
    cleaned = clean_value(value)
    if not cleaned:
        return None
    try:
        return int(cleaned)
    except ValueError:
        return None

def parse_boolean(value):
    """Parse various boolean representations"""
    cleaned = clean_value(value).upper()
    return cleaned in ['TRUE', '1', 'YES']

def parse_rating(value):
    """Parse rating values like '5/5', '4.5/5', '4'"""
    cleaned = clean_value(value)
    if not cleaned:
        return 0
    
    # Handle formats like "5/5", "4.5/5"
    if '/' in cleaned:
        parts = cleaned.split('/')
        try:
            rating = float(parts[0])
            return rating
        except (ValueError, IndexError):
            return 0
    
    try:
        return float(cleaned)
    except ValueError:
        return 0

def parse_date(value):
    """Parse date in DD/MM/YYYY format to YYYY-MM-DD"""
    cleaned = clean_value(value)
    if not cleaned:
        return ''
    
    # Handle DD/MM/YYYY format
    if '/' in cleaned:
        parts = cleaned.split('/')
        if len(parts) == 3:
            try:
                day = parts[0].zfill(2)
                month = parts[1].zfill(2)
                year = parts[2]
                return f"{year}-{month}-{day}"
            except (IndexError, ValueError):
                pass
    
    return cleaned

def parse_gifted(value):
    """Check if book was gifted (has a name in the gifted column)"""
    cleaned = clean_value(value)
    return cleaned != '' and cleaned.lower() != 'false'

def determine_status(read_value, dnf_value):
    """Determine reading status based on Read and DNF columns"""
    read = parse_boolean(read_value)
    dnf = parse_boolean(dnf_value)
    
    if dnf:
        return 'dnf'
    if read:
        return 'read'
    return 'want-to-read'

def csv_to_json(csv_path, output_path):
    """Convert CSV to BookTracker JSON format"""
    books = []
    
    try:
        with open(csv_path, 'r', encoding='utf-8') as csvfile:
            # Use csv.reader to handle proper CSV parsing
            reader = csv.reader(csvfile)
            headers = next(reader)  # Read header row
            
            print(f"CSV Headers: {headers}")
            
            for row_num, row in enumerate(reader, start=2):  # Start at 2 because we skipped header
                if len(row) < len(headers):
                    print(f"Skipping incomplete row {row_num}: {row}")
                    continue
                
                # Map CSV columns to BookTracker format
                book = {
                    'id': generate_id(),
                    'title': clean_value(row[5]) or 'Unknown Title',  # Title
                    'author': clean_value(row[2]) or 'Unknown Author',  # Author
                    'series': clean_value(row[3]) or '',  # Series Title
                    'bookNumber': parse_number(row[4]),  # #In Series
                    'status': determine_status(row[0], row[1]),  # Read, DNF
                    'rating': parse_rating(row[15]),  # Rating
                    'mythicalElement': '',  # Not in CSV
                    'publisher': clean_value(row[16]) or '',  # Publisher
                    'pageCount': parse_number(row[6]),  # Page Count
                    'coverImage': '',  # Not in CSV
                    'edition': clean_value(row[7]) or '',  # Edition
                    
                    # Physical features
                    'digitallySigned': parse_boolean(row[8]),  # Digitally Signed  
                    'signed': parse_boolean(row[9]),  # Signed
                    'sprayedEdges': parse_boolean(row[10]),  # Sprayed Edges
                    'hiddenCover': parse_boolean(row[11]),  # Hidden Cover
                    'reversibleDustJacket': parse_boolean(row[12]),  # Reversable Dust Jacket
                    
                    # Reading dates
                    'startedReading': parse_date(row[13]),  # Started Reading
                    'finishedReading': parse_date(row[14]),  # Finished Reading
                    
                    # Other fields
                    'gifted': parse_gifted(row[17]),  # Gifted
                    'notes': '',  # Not in CSV
                    'dateAdded': datetime.now().isoformat()
                }
                
                books.append(book)
        
        # Write to JSON file
        with open(output_path, 'w', encoding='utf-8') as jsonfile:
            json.dump(books, jsonfile, indent=2, ensure_ascii=False)
        
        print(f"\nSuccessfully converted {len(books)} books to {output_path}")
        return books
        
    except Exception as e:
        print(f"Error converting CSV to JSON: {e}")
        raise

def main():
    csv_path = '/Users/crazybuy/Downloads/Copy of Booklist - Sheet1.csv'
    output_path = '/Users/crazybuy/Book Website/imported-books.json'
    
    try:
        books = csv_to_json(csv_path, output_path)
        
        print(f"\nConversion completed successfully!")
        print(f"Total books converted: {len(books)}")
        print(f"Output file: {output_path}")
        
        # Show some sample data
        print(f"\nSample books:")
        for i, book in enumerate(books[:3]):
            print(f"\n{i+1}. {book['title']} by {book['author']}")
            print(f"   Status: {book['status']}, Rating: {book['rating']}, Pages: {book['pageCount']}")
            
            physical_features = []
            if book['digitallySigned']:
                physical_features.append('Digitally Signed')
            if book['signed']:
                physical_features.append('Signed')
            if book['sprayedEdges']:
                physical_features.append('Sprayed Edges')
            if book['hiddenCover']:
                physical_features.append('Hidden Cover')
            if book['reversibleDustJacket']:
                physical_features.append('Reversible Dust Jacket')
            
            print(f"   Physical features: {', '.join(physical_features) if physical_features else 'None'}")
        
        # Show status summary
        status_counts = {}
        for book in books:
            status = book['status']
            status_counts[status] = status_counts.get(status, 0) + 1
        
        print(f"\nStatus Summary:")
        for status, count in status_counts.items():
            print(f"   {status}: {count}")
            
    except Exception as e:
        print(f"Failed to convert CSV: {e}")
        return 1
    
    return 0

if __name__ == '__main__':
    exit(main())
